### PR TITLE
[runtime]read hdfs file in runtime

### PR DESCRIPTION
**Pull request by zkmn73**
Change-Id: I4dcbbb7fdfe543ea1af67113b79d880cfa3fe411

## Why are these changes needed?
we want to read hdfs files in runtime.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/runtime_env/packaging.py [hunk 1]
```diff
@@ -167,10 +169,20 @@ def parse_uri(pkg_uri: str) -> Tuple[Protocol, str]:
             )
             -> ("gs",
             "gs_public-runtime-env-test_test_module.zip")
+    For HDFS URIs, the path will have '/' replaced with '_'. The package name
+    will be the adjusted path with 'gs_' prepended.
```
Comment (shrekris-anyscale) [inline python/ray/_private/runtime_env/packaging.py:173]:
```suggestion
    will be the adjusted path with 'hdfs_' prepended.
```

**Review by shrekris-anyscale — CHANGES_REQUESTED** [review]
Nice work! The change mostly looks good to me. @simon-mo Do you think we should add an integration test?

File: python/ray/_private/runtime_env/packaging.py [hunk 2]
```diff
@@ -524,10 +543,12 @@ def download_and_unpack_package(
                             "You must `pip install smart_open` "
                             f"to fetch {protocol.value.upper()} URIs."
                         )
-
-                with open(pkg_uri, "rb", transport_params=tp) as package_zip:
-                    with open(pkg_file, "wb") as fin:
-                        fin.write(package_zip.read())
+                if protocol == Protocol.HDFS:
+                    subprocess.check_call(["hdfs", "dfs", "-get", pkg_uri, pkg_file])
```
Comment (simon-mo) [inline python/ray/_private/runtime_env/packaging.py:547]: smart_open already support hdfs https://github.com/RaRe-Technologies/smart_open

**Review by architkulkarni — CHANGES_REQUESTED** [review]
I agree with Simon's comment, can we just use `smart_open` here as we do for the other URIs?

**Pull request closed without merging**